docs/_posts/2019-02-19-saving-state.md

```diff
@@ -0,0 +1,109 @@
+---
+layout: post
+title: "💾 Persisting browser session state"
+date: 2019-02-18 14:35:00 +0200
+author: sebastian
+---
+
+Losing open tabs in a browser can be a painful experience for the user. By itself [SessionManager](https://mozac.org/api/mozilla.components.browser.session/-session-manager/) keeps its state only in memory. This means that something as simple as switching to a different app can cause Android to [kill the browser app's process and reclaim its resources](https://developer.android.com/guide/components/activities/process-lifecycle). The result of that: The next time the user switches back to the browser app they start with a fresh browser without any tabs open.
+
+## Saving state
+
+[Mozilla's Android Components](https://mozac.org/) come with two implementations of session storage and helpers to write your own easily.
+
+### SessionStorage
+
+The [SessionStorage](https://mozac.org/api/mozilla.components.browser.session.storage/-session-storage/) class that comes with the [browser-session](https://github.com/mozilla-mobile/android-components/tree/master/components/browser/session) component saves the state as a file on disk (using [AtomicFile](https://developer.android.com/reference/android/util/AtomicFile) under the hood). It can be used for a browser that wants to have a single state that gets saved and restored (like Fennec or Chrome).
+
+```kotlin
+val sessionStorage SessionStorage(applicationContext, engine)
+
+val sessionManager = sessionManager(engine).apply {
+    sessionStorage.restore()?.let { snapshot -> restore(snapshot) }
+}
+```
+
+ℹ️ Since restoring the last state requires a disk read, it is recommended to perform it off the main thread. This requires the app to gracefully handle the situation where the app starts without any sessions at first. [SessionManager](https://mozac.org/api/mozilla.components.browser.session/-session-manager/) will invoke [onSessionsRestored()](https://mozac.org/api/mozilla.components.browser.session/-session-manager/-observer/on-sessions-restored.html) on a registered [SessionManager.Observer](https://mozac.org/api/mozilla.components.browser.session/-session-manager/-observer/) after restoring has completed.
```

Comment:
> gracefully handle the situation where the app starts without any sessions at first

Should we add a code snippet here from r-b or the sample app where we handle that?

Reply:
That's something we still need to to in RB. It's synchronous there currently. Now that we can handle "no sessions" it should be straight-forward to make this asynchronous.